### PR TITLE
Added support for 'id' table field attribute

### PR DESCRIPTION
**Pull request by Eshva**
- 'id' attribute now can be specified and used by FlatSharp.
- 'id' attribute allowed only of table fields. It's not allowed on struct fields.
- All or none of fields should have - 'id' attribute set.
- 'id' attribute value can not be negative.
- 'id' attribute should have value if present.

### CODE REVIEW

File: src/FlatSharp.Compiler/MetadataHelpers.cs [hunk 1]
```diff
@@ -29,40 +27,45 @@ public static class MetadataHelpers
...
                 key,
                 bool.TryParse,
                 true,
                 false);
         }
 
+        public static bool TryParseIntegerMetadata(this IDictionary<string, string> metadata, string key, out int value)
+        {
+            value = ParseMetadata(metadata, key, int.TryParse, DefaultValueIfPresent, DefaultValueIfNotPresent);
+            return value >= 0;
```
Comment (jamescourtney) [inline src/FlatSharp.Compiler/MetadataHelpers.cs:66]: Given  that it must be non-negative to be valid, it seems like this should be a UInt instead of int, right?

File: src/FlatSharp.sln.DotSettings [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_AFTER_INVOCATION_LPAR/@EntryValue">True</s:Boolean>
+	<s:Int64 x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/BLANK_LINES_AFTER_MULTILINE_STATEMENTS/@EntryValue">1</s:Int64>
```
Comment (jamescourtney) [inline src/FlatSharp.sln.DotSettings:3]: Can you remove this file? I'm not a resharper adherent!

File: src/FlatSharp.Compiler/Visitors/FieldVisitor.cs [hunk 3]
```diff
@@ -33,55 +33,30 @@ public override FieldDefinition VisitField_decl([NotNull] FlatBuffersParser.Fiel
         {
             this.definition.Name = context.IDENT().GetText();
 
-            ErrorContext.Current.WithScope(this.definition.Name, () =>
```
Comment (jamescourtney) [inline src/FlatSharp.Compiler/Visitors/FieldVisitor.cs:36]: Please keep the "this."; it's helpful to identify what's an instance field vs static field.

File: src/FlatSharp.Compiler/Visitors/FieldVisitor.cs [hunk 4]
```diff
@@ -102,7 +77,69 @@ public override FieldDefinition VisitField_decl([NotNull] FlatBuffersParser.Fiel
...
 
-            return this.definition;
+            return definition;
+        }
+
+        private void ParseIdMetadata(IDictionary<string, string> metadata)
+        {
+            if (!metadata.TryParseIntegerMetadata("id", out var index))
+            {
+                if (index == MetadataHelpers.DefaultValueIfPresent)
```
Comment (jamescourtney) [inline src/FlatSharp.Compiler/Visitors/FieldVisitor.cs:87]: Not sure I like the naming here. Or -- I don't like the "defaultvalueifpresent" being on the MetaDataHelpers class. It has a specific meaning in this context, so I think we should define it in this class.

Alternatively, I think the semantic you are really trying to capture here is nullable. Why not have TryParseIntegerMetadata return a nullable int?

Comment (jamescourtney) [reply src/FlatSharp.Compiler/Visitors/FieldVisitor.cs:87]: Also, I wonder how useful this validation is. The compiler is supposed to be at least somewhat dumb. The FlatSharp library (ie, the thing that actually generates the interesting code) will throw a fit if there are two properties with the same index, so we could just let people define whatever values they want here.

File: src/FlatSharp.Compiler/Visitors/FieldVisitor.cs [hunk 5]
```diff
@@ -102,7 +77,69 @@ public override FieldDefinition VisitField_decl([NotNull] FlatBuffersParser.Fiel
...
+
+            definition.Index = index;
+            definition.IsIndexSetManually = true;
+        }
+
+        private void SetFbsFieldType(FlatBuffersParser.Field_declContext context, Dictionary<string, string> metadata)
+        {
+            string fbsFieldType = context.type().GetText();
+
+            definition.VectorType = VectorType.None;
```
Comment (jamescourtney) [inline src/FlatSharp.Compiler/Visitors/FieldVisitor.cs:109]: I think you have a resharper rule to remove "this.".

File: src/FlatSharpTests/FlatSharpCompiler/TableMemberTests.cs [hunk 6]
```diff
@@ -90,7 +90,89 @@ public class TableMemberTests
         [TestMethod]
         public void TableMember_string() => this.RunCompoundTest<string>("string");
 
-        private void RunCompoundTestWithDefaultValue_Bool(string fbsType)
+        [TestMethod]
+        public void TableMember_have_id()
```
Comment (jamescourtney) [inline src/FlatSharpTests/FlatSharpCompiler/TableMemberTests.cs:94]: TableMember_With_Id?

Comment (jamescourtney) [reply src/FlatSharpTests/FlatSharpCompiler/TableMemberTests.cs:94]: Also, thanks for adding a good set of tests!

File: src/FlatSharp.Compiler/TypeDefinitions/FieldDefinition.cs [hunk 7]
```diff
@@ -46,81 +48,6 @@ internal class FieldDefinition
 
         public bool SharedString { get; set; }
 
-        public string GetClrTypeName(BaseSchemaMember baseMember)
```
Comment (jamescourtney) [inline src/FlatSharp.Compiler/TypeDefinitions/FieldDefinition.cs:49]: It's difficult for me to tell what's changed here since methods have moved around. I'm not opposed to reordering anything, but it'd be nice to do that in separate PRs.

**Comment by jamescourtney** [comment]
Hey -- thank you for your contribution! I apologize for my delay in reviewing.

I've added a few comments. I'm not finished reviewing, but I think this set of changes on whole looks fine. There are a couple of files I need to look at in more depth, but the tests all passing is a good sign!

**Comment by Eshva** [comment]
Namaste!
Sorry for my delay, I was busy with my job. I try to find time this weekend to fix your comments.

**Review by Eshva — COMMENTED** [review]
James I will rewrite my changes in another branch and submit it in an another review. But in this are my answers to your comments.

**Comment by Eshva** [comment]
See #85

**Comment by Eshva** [comment]
May be you should decline this PR?
